pkg/platform/api/api.go

```diff
@@ -0,0 +1,211 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/bottlerocket-os/bottlerocket-update-operator/pkg/logging"
+
+	"github.com/pkg/errors"
+)
+
+const bottlerocketAPISock = "/run/api.sock"
+const minimumRequiredOSVer = "0.3.4"
+
+type updateState string
+
+const (
+	Idle      updateState = "Idle"
+	Available updateState = "Available"
+	Staged    updateState = "Staged"
+	Ready     updateState = "Ready"
+)
+
+type UpdateImage struct {
+	Arch    string `json:"arch"`
+	Version string `json:"version"`
+	Variant string `json:"variant"`
+}
+
+func (ui *UpdateImage) Identifier() interface{} {
+	return ui.Version
+}
+
+type StagedImage struct {
+	Image      UpdateImage `json:"image"`
+	NextToBoot bool        `json:"next_to_boot"`
+}
+
+type updateCommand string
+
+const (
+	refresh    updateCommand = "refresh"
+	prepare    updateCommand = "prepare"
+	activate   updateCommand = "activate"
+	deactivate updateCommand = "deactivate"
+)
+
+type commandStatus string
+
+const (
+	Success commandStatus = "Success"
+	Failed  commandStatus = "Failed"
+	Unknown commandStatus = "Unknown"
+)
+
+type CommandResult struct {
+	CmdType    updateCommand `json:"cmd_type"`
+	CmdStatus  commandStatus `json:"cmd_status"`
+	Timestamp  string        `json:"timestamp"`
+	ExitStatus *int32        `json:"exit_status"`
+	Stderr     *string       `json:"stderr"`
+}
+
+type UpdateStatus struct {
+	UpdateState       updateState    `json:"update_state"`
+	AvailableUpdates  []string       `json:"available_updates"`
+	ChosenUpdate      *UpdateImage   `json:"chosen_update"`
+	ActivePartition   *StagedImage   `json:"active_partition"`
+	StagingPartition  *StagedImage   `json:"staging_partition"`
+	MostRecentCommand *CommandResult `json:"most_recent_command"`
+}
+
+type apiClient struct {
+	log        logging.Logger
+	httpClient *http.Client
+}
+
+func newAPIClient() *apiClient {
+	return &apiClient{log: logging.New("api"), httpClient: &http.Client{
+		Transport: &http.Transport{
+			DialContext: func(ctx context.Context, _, _ string) (net.Conn, error) {
+				dialer := net.Dialer{}
+				return dialer.DialContext(ctx, "unix", bottlerocketAPISock)
+			},
+		},
+		// Set a 10 second timeout for all requests
+		Timeout: 10 * time.Second,
+	},
+	}
+}
+
+func (c *apiClient) do(req *http.Request) (*http.Response, error) {
+	var response *http.Response
+	const maxAttempts = 5
+	attempts := 0
+	// Retry up to 5 times in case the Update API is busy; Waiting 10 seconds between each attempt.
+	for ; attempts < maxAttempts; attempts++ {
+		var err error
+		response, err = c.httpClient.Do(req)
+		if err != nil {
+			return nil, errors.Errorf("error when sending request to Bottlerocket API")
```

Comment:
nit: please wrap this error to avoid masking errors

```suggestion
			return nil, errors.Wrapf(err, "update API request error")
```